### PR TITLE
Update Priority Queue to make profiles faster

### LINKED ISSUE #700: get profiling back to perceived "magical"
Profiling has gotten much slower since May. I think there are some easy solutions to this.

1. it would be nice if we re-ordered the queue a bit. Right now, we enqueue a bunch of stuff that the user doesn’t need to see immediately (such as top ns). I’m noticing that profiling feels a lot. slower with the new numeric histogram rug plot. It would feel faster if that was the lowest-priority item in the priority queue.
So I’m thinking of the following priorities:
  - `SummaryProfile` – anything that is directly visible in the profiling, e.g. the initial histograms, null counts, cardinality
  - `EssentialProfileDetails`  – top Ns, timestamp viz
  - `DeeperProfileDetails` – things like the rug plot, which seem to take a bit longer to run.

2. Along with that, we could prefetch a detail if the user mouses over the profile. This would amount to re-prioritizing the details to be "UserRequestedProfile", which should be above `SummaryProfile` here. @domoritz suggests also taking a look at [premonish](https://github.com/mathisonian/premonish), which uses some simple heuristics to determine if a user's mouse is moving toward a DOM element. This would buy us another 200ms to reorder the priority queue, which would likely make it so that the detail views are already calculated.

### PR DESCRIPTION
**Pull request by djbarnwal**
fixes #700

### CODE REVIEW

File: src/lib/components/column-profile/ColumnProfileDetails.svelte [hunk 1]
```diff
@@ -16,11 +19,23 @@
...
+
+  // Make sure priority is updated in case the profile is already opened
+  afterUpdate(async () => {
+    if (active) {
+      dataModelerService.dispatch("updateColumnPriority", [entityId, name]);
+    }
+  });
+  // $: if (active) {
+  //   dataModelerService.dispatch("updateColumnPriority", [entityId, name]);
+  // }
```
Comment (djbarnwal) [inline src/lib/components/column-profile/ColumnProfileDetails.svelte:38]: @AdityaHegde @hamilton Is there a better way to achieve the same with minimal re-renders? Using `$: if(active)` does not execute if the profile is already active

File: src/common/data-modeler-service/ApplicationActions.ts [hunk 2]
```diff
@@ -73,6 +111,21 @@ export class ApplicationActions extends DataModelerActions {
 
     this.databaseActionQueue.clearQueue(entityId);
 
+    if (entityType === EntityType.Model || entityType === EntityType.Table) {
+      // Clear existing profile actions in queue
+      const columns =
```
Comment (AdityaHegde) [inline src/common/data-modeler-service/ApplicationActions.ts:116]: There are multiple instances of this piece. Please move it to a separate method.

File: src/lib/components/column-profile/ColumnProfileDetails.svelte (hunk 1 above)
Comment (AdityaHegde) [reply src/lib/components/column-profile/ColumnProfileDetails.svelte:41]: We should call this where active is actually changed.

File: src/common/data-modeler-service/ApplicationActions.ts [hunk 3]
```diff
@@ -27,17 +33,49 @@ export class ApplicationActions extends DataModelerActions {
...
+            getProfilePriority(
+              DatabaseActionQueuePriority.InactiveModelProfile,
+              DatabaseProfilesFieldPriority.NonFocused,
+              ProfileMetadataPriorityMap[priority]
+            )
+          );
+        });
+      });
     }
     this.dataModelerStateService.dispatch("setActiveAsset", [
```
Comment (AdityaHegde) [inline src/common/data-modeler-service/ApplicationActions.ts:55]: Where are we setting all column profiles as active when new active asset is a model?

File: src/lib/components/column-profile/ColumnProfileDetails.svelte (hunk 1 above)
Comment (hamilton) [reply src/lib/components/column-profile/ColumnProfileDetails.svelte:41]: agreed, would be better to have this call not as a reactive statement but where there is a click callback

Comment (djbarnwal) [reply src/lib/components/column-profile/ColumnProfileDetails.svelte:41]: The issue with having it in the callback is that when a profile is already opened and the user changes the query, the `active` callback is not executed thus no action is dispatched to prioritize that column

**Comment by djbarnwal** [comment]
Adding @magorlick for a review of the UX

**Review by AdityaHegde — APPROVED** [review]

**Comment by magorlick** [comment]
When I first started this branch I was immediately impressed with how fast the profiling felt. I did an audible gasp. Great work. :)  I almost didn't believe it so I gave it a good hard kick to see where it isn't working. I noticed a few things.

Sometimes when I start this branch of Rill Developer it appears I have lost all of my sources and models.  Is there any way to prevent this?  Is it perhaps tied to waiting to load until profiling is done?
Kapture 2022-08-12 at 17 54 59

When I am navigating between models I am noticing that the model takes a beat before loading.  Is this possibly related?
Kapture 2022-08-12 at 18 01 03

Did you implement this to run profiles above the fold before those below? Could we do them in sort-order almost like a progress bar working down the page? Sometimes it feels off to run them out of order and I am looking at the idle indicator wondering if something is wrong.

The buttons in the upper right hand corner flicker when I am editing.  The sources in the right hand nav also disappear. They should always be there with the text updating.
Kapture 2022-08-12 at 18 10 45

**Comment by djbarnwal** [comment]
@magorlick 

1. Regarding sources not loading on branch change, I won't pay much attention to that. Once this gets merged in it should be a non issue
2. I am unable to reproduce the lag you are showing while switching the models. Maybe you can try restarting the server and see if it's still an issue
3. As of now we are not prioritizing profiles based on their position. The priority is based on [this proposal](https://www.notion.so/rilldata/Reordering-Rill-Developer-s-Priority-Queue-ee8d8cb0382a4df9aabaa8c69011d90b). Although @hamilton and I were discussing virtualization of the profiles (just like the tables). It would make sense to incorporate the request change during the virtualization process
4. I can reproduce the flickering issue on `main` as well. Would be good to have an issue for it.

**Comment by magorlick** [comment]
> @magorlick
> 
> 1. Regarding sources not loading on branch change, I won't pay much attention to that as of now. Once this gets merged in it should be a non issue
> 2. I am unable to reproduce the lag you are showing while switching the models. Maybe you can try restarting the server and see if it's still an issue

These two issues are present in the current release, I didn't notice them before because 1) I was kicking a little harder with a massive data set and 2) the dashboard assets are loading faster than sources and models in this branch so the latency is more obvious and feels a little broken.  The solution is an application loading screen that is out of scope for this PR.  I am filing another issue and this is ok for the next release.

> 3. As of now we are not prioritizing profiles based on their position. The priority is based on [this proposal](https://www.notion.so/rilldata/Reordering-Rill-Developer-s-Priority-Queue-ee8d8cb0382a4df9aabaa8c69011d90b). Although @hamilton and I were discussing virtualization of the profiles (just like the tables). It would make sense to incorporate the requested change during the virtualization process

This sounds good, we can wait until we virtualize the list.  I think this will capture a lot of magic! I had many columns in my data set and it didn't feel quite as snappy when it was loading below the fold.

> 4. I can reproduce the flickering issue on `main` as well. Would be good to have an issue for it.

I see that too.  Filing another issue to cover the button flicker as it is not related to these changes.

Together, we have some follow up work to do but this is a nice step in the right direction and I approve the experience.  Thank you for your work on this!

**Review by magorlick — APPROVED** [review]

**Pull request merged**